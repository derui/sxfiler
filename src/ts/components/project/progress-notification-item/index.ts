export * from "./progress-notification-item";
