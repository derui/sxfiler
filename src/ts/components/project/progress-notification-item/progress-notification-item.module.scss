@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.root {
  composes: listItem-base from "../../ui/list-item/list-item.module.scss";

  background-color: $sol_base02;
}

.processLabel {
  flex: 0 0 20%;
  padding: 0 $smallSpace;
  color: $sol_base2;

  border-right: $smallSpace solid $sol_base3;
  max-width: 20%;

  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.progressBar {
  flex: 1 1 auto;
  height: $smallSpace;

  margin: auto $smallSpace;

  background-color: $sol_base2;

  border-radius: $smallSpace;
}

.progressIndicator {
  height: 100%;
  background-color: $sol_blue;

  transition: width 100ms ease-out;

  border-radius: $smallSpace;
}
