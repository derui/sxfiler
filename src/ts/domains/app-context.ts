import { UIContext } from "@/types/ui-context";

export type AppContext = {
  readonly current: UIContext;
  readonly subContexts: UIContext[];
};

/**
   create AppContext
 */
export function createAppContext({
  current,
  subContexts,
}: {
  current: UIContext;
  subContexts?: UIContext[];
}): AppContext {
  return {
    current,
    subContexts: Array.from(subContexts || []),
  };
}

/**
   change context
 */
export function changeCurrent(context: UIContext) {
  return (state: AppContext) =>
    createAppContext({
      current: context,
      subContexts: state.subContexts,
    });
}

/**
   Add a context to AppContext
 */
export function addSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.add(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}

/**
   remove the context from AppContext
 */
export function removeSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.delete(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}
