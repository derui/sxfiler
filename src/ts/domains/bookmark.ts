/**
   Type of bookmark
 */
export type Bookmark = {
  readonly id: string;
  readonly path: string;
  readonly order: number;
};

type FactoryArg = Pick<Bookmark, keyof Bookmark>;

/**
 * Create bookmark from factory argument
 */
export const createBookmark = function createBookmark({ id, path, order }: FactoryArg): Bookmark {
  return {
    id,
    path,
    order,
  };
};
