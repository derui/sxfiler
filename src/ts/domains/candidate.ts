export type CandidateObject = {
  readonly id: string;
  readonly value: string;
  readonly start: number;
  readonly length: number;
};

export type Candidate = CandidateObject & {
  /**
   * Get splitted candidate as 3 pieces by matched string that
   * has the one of before start of match, matched, and after matched.
   */
  splitByInput(): [string, string, string];

  /**
   * Get plain object.
   */
  plain(): CandidateObject;
};

type FactoryArg = Pick<CandidateObject, "id" | "value"> & Partial<Pick<CandidateObject, "start" | "length">>;

/**
 * Create candidate from candidate object.
 */
export const createCandidate = ({ id, value, start = 0, length = 0 }: FactoryArg): Candidate => {
  return {
    id,
    value,
    start: Math.max(0, Math.min(value.length - 1, start)),
    length: Math.max(0, Math.min(value.length, length)),
    splitByInput() {
      const beforeMatched = value.substring(0, start);
      const matched = value.substring(start, start + length);
      const afterMatched = value.substring(start + length);

      return [beforeMatched, matched, afterMatched];
    },

    plain() {
      return { id: this.id, value: this.value, start: this.start, length: this.length };
    },
  };
};
