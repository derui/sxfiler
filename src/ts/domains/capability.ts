// Capability with method
export type Capability = CapabilityObject & {
  /** Return new capability that allows to write */
  allowToWrite(): Capability;

  /** Return new capability that allows to read */
  allowToRead(): Capability;

  /** Return new capability that allows to execute */
  allowToExecute(): Capability;

  /** Return new capability that disallows to write */
  disallowToWrite(): Capability;

  /** Return new capability that disallows to read */
  disallowToRead(): Capability;

  /** Return new capability that disallows to execute */
  disallowToExecute(): Capability;

  /** Return plain CapabilityObject */
  plain(): CapabilityObject;
};

// Capability of a target
export type CapabilityObject = {
  readonly writable: boolean;
  readonly readable: boolean;
  readonly executable: boolean;
};

type CreateCapabilityArg = {
  writable: boolean;
  readable: boolean;
  executable: boolean;
};

function plain(this: Capability): CapabilityObject {
  return { writable: this.writable, readable: this.readable, executable: this.executable };
}

/**
   Create capability from argument
 */
export const createCapability = ({ writable, readable, executable }: CreateCapabilityArg): Capability => {
  return {
    writable,
    readable,
    executable,

    plain,

    /** Return new capability that allows to write */
    allowToWrite() {
      return { ...this, writable: true };
    },

    /** Return new capability that allows to read */
    allowToRead() {
      return { ...this, readable: true };
    },

    /** Return new capability that allows to execute */
    allowToExecute() {
      return { ...this, executable: true };
    },

    /** Return new capability that disallows to write */
    disallowToWrite() {
      return { ...this, writable: false };
    },

    /** Return new capability that disallows to read */
    disallowToRead() {
      return { ...this, readable: false };
    },

    /** Return new capability that disallows to execute */
    disallowToExecute() {
      return { ...this, executable: false };
    },
  };
};

/**
   Return new capability that have not any work.
 */
export const emptyCapability = (): Capability =>
  createCapability({ writable: false, readable: false, executable: false });

/**
   Return new capability that have all capability on all roles.
 */
export const fullCapability = (): Capability => createCapability({ writable: true, readable: true, executable: true });
