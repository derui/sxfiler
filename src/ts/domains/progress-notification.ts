// type of notification

export type Body = {
  readonly process: string;
  readonly current: number;
  readonly target: number;
};

export type ProgressNotification = {
  readonly id: string;
  readonly body: Body;
};

/**
 * factory function to create progress notification
 * @param id
 * @param level
 * @param progress
 */
export const createProgress = (
  id: string,
  progress: { process: string; current: number; target: number }
): ProgressNotification => {
  return {
    id,
    body: { ...progress },
  };
};
