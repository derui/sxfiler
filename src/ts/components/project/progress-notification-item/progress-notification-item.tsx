import * as React from "react";
import { ProgressBody } from "../../../domains/notification";
import * as Element from "../../ui/element/element";
import * as ListItem from "../../ui/list-item/list-item";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./progress-notification-item.module.scss");

// make a progress bar
function makeProgressBar(ratio: number) {
  const style = {
    width: `${ratio}%`,
  };

  return (
    <div key="progress-bar" className={styles.progressBar}>
      <div className={styles.progressIndicator} style={style} />
    </div>
  );
}

const Span = Element.createComponent({ tagName: "span" });

export interface Props {
  body: ProgressBody;
}

export class Component extends React.PureComponent<Props> {
  constructor(props: Props) {
    super(props);
  }

  public render() {
    const { body } = this.props;
    const ratio = Math.min(100, (body.current / body.target) * 100);

    return (
      <ListItem.Component className={styles.root}>
        <Span className={styles.processLabel}>{body.process}</Span>
        {makeProgressBar(ratio)}
      </ListItem.Component>
    );
  }
}
