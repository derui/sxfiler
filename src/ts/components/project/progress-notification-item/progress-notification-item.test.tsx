import * as React from "react";
import renderer from "react-test-renderer";

import * as Notification from "../../../domains/progress-notification";

import { Component as T } from "./progress-notification-item";

describe("Project", () => {
  describe("Progress Notification Item", () => {
    it("should print correctly", () => {
      const notification = Notification.createProgress("id", {
        process: "target process",
        current: 10,
        target: 100,
      });
      const tree = renderer.create(<T body={notification.body} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should hide when current value is equvalent to target value", () => {
      const notification = Notification.createProgress("id", {
        process: "target process",
        current: 101,
        target: 100,
      });
      const tree = renderer.create(<T body={notification.body} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should not over 100% of width", () => {
      const notification = Notification.createProgress("id", {
        process: "target process",
        current: 101,
        target: 100,
      });
      const tree = renderer.create(<T body={notification.body} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
