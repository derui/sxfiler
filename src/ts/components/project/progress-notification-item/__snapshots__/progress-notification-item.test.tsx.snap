// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Progress Notification Item should hide when current value is equvalent to target value 1`] = `
<div
  aria-selected={false}
  className="sc-bdVaJa gneklD"
  role="listitem"
>
  <span
    className="sc-bwzfXH cJKlyb"
  >
    target process
  </span>
  <div
    className="sc-htpNat hkuyqD"
  >
    <div
      className="sc-bxivhb kUGzNj"
    />
  </div>
</div>
`;

exports[`Project Progress Notification Item should not over 100% of width 1`] = `
<div
  aria-selected={false}
  className="sc-bdVaJa gneklD"
  role="listitem"
>
  <span
    className="sc-bwzfXH cJKlyb"
  >
    target process
  </span>
  <div
    className="sc-htpNat hkuyqD"
  >
    <div
      className="sc-bxivhb kUGzNj"
    />
  </div>
</div>
`;

exports[`Project Progress Notification Item should print correctly 1`] = `
<div
  aria-selected={false}
  className="sc-bdVaJa gneklD"
  role="listitem"
>
  <span
    className="sc-bwzfXH cJKlyb"
  >
    target process
  </span>
  <div
    className="sc-htpNat hkuyqD"
  >
    <div
      className="sc-bxivhb MqTMw"
    />
  </div>
</div>
`;
